source/lac/sparsity_tools.cc

```diff
@@ -358,12 +361,28 @@ namespace SparsityTools
     {
       AssertDimension (connectivity.n_rows(), connectivity.n_cols());
       AssertDimension (connectivity.n_rows(), renumbering.size());
+      Assert (connectivity.row_index_set().size() == 0 ||
+              connectivity.row_index_set().size() == connectivity.n_rows(),
+              ExcMessage("Only valid for sparsity patterns which store all rows."));
 
       std::vector<types::global_dof_index> touched_nodes(connectivity.n_rows(),
                                                          numbers::invalid_dof_index);
+      std::vector<unsigned int> row_lengths(connectivity.n_rows());
+      std::vector<unsigned int> n_remaining_neighbors(connectivity.n_rows());
       std::set<types::global_dof_index> current_neighbors;
       std::vector<std::vector<types::global_dof_index> > groups;
 
+      // First collect the number of neighbors for each node. We use this
+      // field to find next nodes with the minimum number of non-touched
+      // neighbors, so we will count down on this field. We also cache the row
+      // lengths because we need this data frequently and getting it from the
+      // sparsity pattern is expensive.
+      for (types::global_dof_index row=0; row<connectivity.n_rows(); ++row)
+        {
+          n_remaining_neighbors[row] = row_lengths[row] = connectivity.row_length(row);
```

Comment:
I think you can write this more efficiently (and more easy to read) if you do only one assignment here and after the loop copy the one array into the other (preferably into the one that you're going to continue modifying). This also allows you to declare the array you're going to copy into below this loop and simply use copy assignment.